ml/tests/test_dataset.py

```
@@ -266,3 +267,161 @@ def test_verify_rejects_a_format_the_training_decoder_cannot_read(tmp_path):
+
+
+@requires_tensorflow
+def test_unshuffled_order_matches_the_entries(cache_entries, cache_config):
+    """Without shuffling, the cache reorders nothing."""
+    from src.dataset import build_dataset
+
+    ds = build_dataset(cache_entries, cache_config)
+    assert _means(ds) == _means(ds), "an unshuffled dataset changed order between epochs"
+
+
+def _means(ds):
```

Ruff (0.16.3)

[warning] Missing return type annotation for private function `_means`

(ANN202)